libsnark/zk_proof_systems/plonk/plonk.tcc

```diff
@@ -0,0 +1,447 @@
+//                      [a0, a1, ..., a_{n-1}] is a vector
+//                      representing the coefficients of the i-th
+//                      Lagrange polynomial L_i(x) =
+//                      a0+a1x+a2x^2+..+a_{n-1}x^{n-1}
+//                      s.t. L_i(x=omega_i)=1 and
+//                      L_i(x\neq{omega_i)})=0
+//
+// Note: uses libfqfft iFFT for the interpolation
+//
+template<typename FieldT>
+void plonk_compute_lagrange_basis(
+    size_t npoints, std::vector<polynomial<FieldT>> &L)
+{
+    assert(L.size() != 0);
+    assert(L.size() == L[0].size());
+    assert(L.size() == npoints);
+
+    std::shared_ptr<libfqfft::evaluation_domain<FieldT>> domain =
+        libfqfft::get_evaluation_domain<FieldT>(npoints);
+    for (size_t i = 0; i < npoints; ++i) {
+        polynomial<FieldT> u(npoints, FieldT(0));
+        u[i] = FieldT(1);
+        // compute i-th Lagrange basis vector via inverse FFT
+        domain->iFFT(u);
+        L[i] = u;
+    }
+}
+
+//
+// Interpolate a set of points as a polynomial over Lagrange basis
+//
+// INPUT:
+//
+// - f_points[0..n-1]: a set of points (0,y0), (1,y1),
+//                     ... (n-1,y_{n-1}) s.t. y0=f_points[0],
+//                     y1=f_points[1], ... which we want to interpolate
+//                     as a polynomial
+//
+// - L[0..n-1][0..n-1]: Lagrange basis over the n roots of unity
+//                      omega_0, ..., omega_{n-1} i.e. L[omega_i] =
+//                      [a0, a1, ..., a_{n-1}] is a vector
+//                      representing the coefficients of the i-th
+//                      Lagrange polynomial L_i(x) =
+//                      a0+a1x+a2x^2+..+a_{n-1}x^{n-1}
+//                      s.t. L_i(x=omega_i)=1 and
+//                      L_i(x\neq{omega_i)})=0
+//
+// OUTPUT:
+//
+// - f_poly[0..n-1]: the coefficients [a0, a1, ..., a_{n-1}] of the
+//                   polynomial f(x) interpolating the set of points
+//                   f_points over the Lagrange basis. For example if
+//                   f_poly[0..n-1] = [a0, a1, ..., a_{n-1}] then
+//                   this represents the polynomial f(x) =
+//                   \sum^{n-1}_{i=0} f_vec[i] * L[i] =
+//                   a0+a1x+a1x^2+...+a_{n-1}x^{n-1} such that
+//                   f(omega_i)=f_vec[i].
+//
+template<typename FieldT>
+void plonk_interpolate_over_lagrange_basis(
+    std::vector<FieldT> f_points,
+    polynomial<FieldT> &f_poly,
+    std::vector<polynomial<FieldT>> L)
+{
+    assert(L.size() != 0);
+    assert(L.size() == L[0].size());
+    assert(L.size() == f_points.size());
+
+    size_t num_gates = L.size();
+
+    // the num_gates components of f_poly. each components is
+    // a polynomial that is the multiplication of the i-th element of
+    // f_points to the i-th polynomial in the lagrange basis L[i]:
+    // f_poly_component[i] = f_points[i] * L[i]
+    std::vector<polynomial<FieldT>> f_poly_component(num_gates);
```

Comment:
We could avoid allocating many vectors by creating each component and adding them immediately to `f_poly`.

However, is it not more efficient to simply compute the ifft?

Reply:
Fixed the function to accumulate iteratively the sum, as as to avoid allocating all vectors at the start. Also added a new function which uses `libfqfft iFFT` to perform the same computation.

Now a dilemma arises: is the function `plonk_compute_lagrange_basis` necessary at all? Arguments for "yes" are:

1. it can be used even when the number of gates is not a power of 2, 
2. Lagrange basis interpolation is used everywhere in the original Plonk paper [GWC19]. 

Arguments for "no" are: 

1. The native libfqfft iFFT function is arguably more efficient and less bug prone, so let's only use that (BUT it requires number of gates to be power of 2).

I have left the `plonk_compute_lagrange_basis` function for now until we resolve the dilemma to remain consistent with [GWC19]. Note that it is not used anywhere in the current implementation though.

Reply:
This was briefly discussed in the weekly meeting on 14 June. Decided to only pass the `libfqfft domain` as input to functions instead of all Lagrange basis polynomials, roots of unity, `Zh` polynomial, etc. The latter can be computed from the domain when needed. Agreed to apply this change in the long run. Not applied for the next round of reviews.